Add definitions for grouping media players

klada opened the pull request:
## Proposed change

For multiroom audio systems (Sonos, HEOS, Bluesound, ...) we should provide a common API for joining media players together. This has been discussed in https://github.com/home-assistant/architecture/issues/364 and #32568

As of now this API consists of the following things:

- Generic services defined in the media_player component: `media_player.join`, `media_player.unjoin`.
- A flag `SUPPORT_GROUPING` for media players indicating whether the platform has support for joining players together
- Two new media_player properties:
  - `group_members`: A list of players which are grouped together
  - `support_grouping`: Exposes the `SUPPORT_GROUPING` capability as a media player property
- Two new methods `async_join_players` and `async_unjoin` which need to be defined in media players which actually support grouping.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:

No config changes required.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16442
- Link to developer documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/855

This PR is a preparation for bringing multiroom audio support to the HEOS platform (#32568). That PR is moving forward very slowly though so splitting this commit out should make the review for the integration maintainer easier.

- This PR is related to issue: https://github.com/home-assistant/architecture/issues/364


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

klada committed:
Add definitions for grouping media players

See https://github.com/home-assistant/architecture/issues/364

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -297,6 +304,12 @@ async def async_setup(hass, config):
     )
+    component.async_register_entity_service(
+        SERVICE_JOIN,
+        {vol.Required(ATTR_GROUP_MEMBERS): list},
```
Should there be a way to define a leader ?

klada replied:
@balloob Thank you very much for your input. I was initially thinking about explicitly specifying the leader as well. However @andrewsayre suggested to leave this out (https://github.com/home-assistant/core/pull/32568#pullrequestreview-454203730).

Specifying a leader could make sense for some integrations (HEOS for example), as the leader speaker is static there. However some multiroom audio systems (such as Sonos) have the concept of taking over leadership dynamically, if the leader speaker is removed from an existing group. There *might* also be other multiroom audio systems out there which don't have a leader concept at all, but I am not sure about this.

So the idea of @andrewsayre was to hide the leader concept entirely and let the integration handle the details.

But yeah, in the end we have multiple options here:

1. Hide the leader concept from the service calls and leave it up to the integration

2. Require an explicit leader in the service call

3. Make the leader an optional argument in the service call

I am fine with either approach. Currently I chose **1.** but I'll happily implement one of the other ones. I must admit that I am a little stuck right now because I don't know where I can get proper feedback for such core decisions. The arch issue (home-assistant/architecture/issues/364) does not contain any comments regarding this. How do we want to handle this?

balloob replied:
Hiding the leader concept seems ok 👍

klada replied:
OK, perfect.

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -512,6 +528,22 @@ def shuffle(self):
         return None
 
+    @property
+    def group_candidates(self):
```
I suggest we leave this out for now. It can be considered to be added later.

klada replied:
Sure, I can just leave that out :)

klada replied:
OK, I just dropped the `group_candidates` attribute for now. Let's wait if people actually need it.

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -512,6 +528,22 @@ def shuffle(self):
+        return []
+
+    @property
+    def group_members(self):
```
If the entity supports grouping, we should make sure that we include this as an attribute in `device_attributes`.

klada replied:
Right now `group_members` is always exposed. Do you mean we should only expose it if the device actually supports grouping? Or do you mean we should expose an additional attribute `support_grouping`?

balloob replied:
Ah I see you added it to `ATTR_TO_PROPERTY`. Let's add it manually to `state_attributes` because we want to only add this attribute if we support grouping.

klada replied:
OK, that makes perfectly sense. Fixed.

balloob reviewed (changes requested):
Please add a test to make sure this functionality works.

klada commented:
> Please add a test to make sure this functionality works.

OK, I have added tests for the service calls. We still need to decide how the calls should look like, see my inline comment (https://github.com/home-assistant/core/pull/41193#discussion_r560872378).

balloob commented on homeassistant/components/demo/media_player.py:
```diff
@@ -398,6 +407,18 @@ def set_repeat(self, repeat):
+        self._group_members = [
+            self.entity_id,
+        ] + group_members
+        self.schedule_update_ha_state()
```
You're in async context.

```suggestion
        self.async_write_ha_state()
```

balloob commented on homeassistant/components/demo/media_player.py:
```diff
@@ -398,6 +407,18 @@ def set_repeat(self, repeat):
+    async def async_unjoin(self):
+        """Remove this player from any group."""
+        self._group_members = []
+        self.schedule_update_ha_state()
```
```suggestion
        self.async_write_ha_state()
```

klada replied:
Thanks, the demo player is updated now.

balloob reviewed (approved):
This looks great. Almost there!

klada commented:
I have applied the changes suggested by the review from @balloob. From my side everything is set.

However I do not know about the failing tests. The failing test (`test_discover_lights`) does not have anything to do with the stuff I did :smile:

balloob commented:
Looks all good. Failing CI can be fixed by rebasing the PR. Can be merged when the docs PR is added 🎉

klada commented:
Doc are ready: https://github.com/home-assistant/home-assistant.io/pull/16442

Now I'll take care of implementing it in HEOS.

frenck committed:
Fix Google Assistant tests

frenck commented:
There was a merge conflict, I've rebased this PR and adjusted a failing test in Google Assistant.

MartinHjelmare commented on homeassistant/components/demo/media_player.py:
```diff
@@ -398,6 +407,18 @@ def set_repeat(self, repeat):
         self._repeat = repeat
         self.schedule_update_ha_state()
 
+    async def async_join(self, group_members):
```
These methods also need to be defined in the base media player entity class. See other service methods for examples.

klada replied:
@MartinHjelmare Thanks for your suggestion. I have now added the two new methods to the base media player class (with a NotImplemented exception).

I also had to rename `async_join` to `async_join_players`, as it turned out that `async_join` was already in use by some other components and pylint did not like that :)

I have also updated the PR description so this is now ready from my side.

MartinHjelmare commented:
Please also update the PR description with the latest plan.

MartinHjelmare commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -861,6 +887,14 @@ async def async_browse_media(
         """
         raise NotImplementedError()
 
+    async def async_join_players(self, group_members):
```
We normally implement sync versions too of the service handlers and have these async coroutine functions await the sync methods scheduled in the executor with `hass.async_add_executor_job`. Raise NotImplementedError in the sync methods.

klada replied:
@MartinHjelmare Sorry about the squashing. I have now added a sync version of the two methods as well.

MartinHjelmare reviewed (commented):
Please don't squash commits after review has started to make it easier for readers to track changes.

klada committed:
Define sync versions of async_join_players/async_unjoin

MartinHjelmare commented on homeassistant/components/demo/media_player.py:
```diff
@@ -398,6 +407,18 @@ def set_repeat(self, repeat):
+        self._group_members = [
+            self.entity_id,
+        ] + group_members
+        self.async_write_ha_state()
```
In sync functions we need to use our sync api, `self.schedule_update_ha_state`. Or go back to use the coroutine function and async api.

klada replied:
You are right, I have overlooked that. Strangely tests were passing... Anyway, it's fixed now.

klada committed:
Don't use async API in synchronous test methods

MartinHjelmare reviewed (approved):
Looks good!

klada committed:
Fix tests and make pylint happy

The method name `unjoin` is used by another component, so let's use
`unjoin_player` instead.

klada committed:
Fix emulated_hue tests

The new media player entity in the `demo` component requires a tiny
adjustment in the emulated_hue tests.

MartinHjelmare commented on homeassistant/components/media_player/services.yaml:
```diff
@@ -184,3 +184,30 @@ repeat_set:
+    Group players together. Only works on platforms with support for player
+    groups.
+  fields:
+    entity_id:
```
Please use `target:` instead. See other service descriptions for examples.

klada replied:
OK, fixed.

klada committed:
Use "target:" in service description

MartinHjelmare commented on homeassistant/components/media_player/services.yaml:
```diff
@@ -184,3 +184,22 @@ repeat_set:
             - "one"
+
+join:
+  description:
```
Please add a `name` item too.

klada replied:
OK, I have updated the service description.

klada committed:
Also use "name:" in service descriptions

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
Please link a dev docs PR and a user docs PR:

https://developers.home-assistant.io/docs/core/entity/media-player

https://www.home-assistant.io/integrations/media_player/#services

frenck commented:
The one for the end-user documentation is here: home-assistant/home-assistant.io#16442
So that makes the one for the developer documentation missing at this point.

klada commented:
@MartinHjelmare Dev docs PR is ready: https://github.com/home-assistant/developers.home-assistant/pull/855

MartinHjelmare commented:
Thanks!
